### PR TITLE
feat(datastore): add target versioning and discoverable field

### LINKED ISSUE #16: Add "discover" property to target and abandon "scanTargets" config
**What problem does this solve?**

Discovery uses the `scanTargets` configuration to establish the list of targets it will scan for resources. This configuration is disconnected from the actual targets the user might have previously created with apply commands. This means we are asking users to do double bookkeeping, which is error-prone and cumbersome.

**Describe the solution you'd like**:

By adding a `discover` property to a target, a user can immediately, upon creating the target, specify whether they want this target to be included in the resource discovery process. Removing the config option also means that we need to support applying a forma file that contains only targets. Right now if you apply a forma that only contains targets, the metastructure will immediately return because the forma generates no resource updates. Akin to resource update generation, we need to start generating target updates in the metastructure.

The only property that can be changed on targets is the new `discovery` property. If a user changes the config, an error will be returned.

For this to be useful, we also need to add a `formae inventory targets` command so users can view their targets.

**Out of scope**:

If we want to fully honor the `--mode patch` and `--mode reconcile` semantics, we need to generate resource updates to delete resources for targets that were removed in a `reconcile` operation. This is out of scope for now, targets are modify and append-only, making `--mode patch` and `--mode reconcile` effectively behave the same.

### PR DESCRIPTION
**Pull request by browdues**
# Add target management

## Overview
Discovery targets are now managed within forma and the idea of scan targets has been deprecated
- Targets are versioned to track changes over time
- Config-based targets automatically migrate to DB
- Targets default to non-discoverable

### Discovery
- New bool: `discoverable` in target definition
- Target discovery updated via forma files
- Discovery loads targets from database using `LoadDiscoverableTargetsDistinctConfig()`

### Datastore
- Split `StoreTarget()` into `CreateTarget()` and `UpdateTarget()`
- All target queries return latest version
- Both sqlite and pg impelmentations updated

### CLI Display
- Added Target Updates, as shown. These always appear on top of updates:
image

### Usage
- Now possible to create forma file with target only
```pkl
forma {
    new formae.Target {
        label = "aws-prod"
        namespace = "AWS"
        discoverable = true
        config {
            type = "AWS"
            region = "us-west-2"
        }
    }
}
```

### Testing
- Unit tests for target updates + validation
- Enhanced discovery tests with edge case coverage
- Datastore versioning tests
- Renderer tests for update/creation display

## Related Issues
Closes #16

### CODE REVIEW

File: internal/cli/renderer/renderer.go [hunk 1]
```diff
@@ -24,7 +24,7 @@ func RenderSimulation(s *apimodel.Simulation) (string, error) {
 	buf := strings.Builder{}
 	renderHeader := func(cmd apimodel.Command) string { return "Command will" }
 
-	command, err := renderCommand(s.Command, renderHeader, formatSimulatedResourceUpdate)
+	command, err := renderCommand(s.Command, renderHeader, formatSimulatedResourceUpdate, true)
```
Comment (JeroenSoeters) [inline internal/cli/renderer/renderer.go]: Can we use the same pattern as for the resource update and pass in a func that renders the thing? That way we don't have boolean flags as parameters and if-statements. The intent of the previous refactoring was to keep the renderCommand method not simulation-aware.

File: internal/metastructure/datastore/datastore.go [hunk 2]
```diff
@@ -78,10 +78,12 @@ type Datastore interface {
 	LoadStack(stackLabel string) (*pkgmodel.Forma, error)
 	LoadAllStacks() ([]*pkgmodel.Forma, error)
 
-	StoreTarget(target *pkgmodel.Target) (string, error)
+	CreateTarget(target *pkgmodel.Target) (string, error)
+	UpdateTarget(target *pkgmodel.Target) (string, error)
 	LoadTarget(targetLabel string) (*pkgmodel.Target, error)
 	LoadAllTargets() ([]*pkgmodel.Target, error)
 	LoadTargetsByLabels(targetNames []string) ([]*pkgmodel.Target, error)
+	LoadDiscoverableTargetsDistinctConfig() ([]*pkgmodel.Target, error)
```
Comment (JeroenSoeters) [inline internal/metastructure/datastore/datastore.go]: nit: let's just call this `LoadDiscoverableTargets`, it will be implied that this is deduplicated

File: internal/metastructure/discovery/discovery.go [hunk 3]
```diff
@@ -621,3 +647,57 @@ func renderSummary(summary map[string]int) string {
...
+				continue
+			}
+
+			proc.Log().Debug("Created scan_target in database", "target", configTarget.Label)
+			createdCount++
+		}
+	}
+
+	if createdCount > 0 || updatedCount > 0 {
+		proc.Log().Warning("MIGRATION COMPLETE: Please remove scan_targets from your config file and use discoverable targets in forma files instead. In the next release, scan_targets in config will be completely ignored.")
```
Comment (JeroenSoeters) [inline internal/metastructure/discovery/discovery.go]: We can probably be a bit more specific here. What about:

> DEPRECATION WARNING: The 'scanTargets' configuration option is deprecated and will be removed in a future version. 
> 
> Please transition to using the 'discover' flag on individual targets in your Forma files instead.
> 
> For this release, any targets specified in 'scanTargets' will be automatically migrated:
> - If a target already exists in your infrastructure, its 'discover' flag will be set to true
> - If a target doesn't yet exist in your infrastructure, it will be created with 'discover' set to true
> 
> No immediate action is required, but we recommend removing 'scanTargets' from your configuration file and managing discovery through the 'discover' flag on individual targets instead.


We could also print it instead of log it. Maybe we can also throw in a link to the new docs.

File: internal/metastructure/forma_persister/forma_command_persister.go [hunk 4]
```diff
@@ -25,6 +25,11 @@ import (
 	"github.com/platform-engineering-labs/formae/pkg/plugin/resource"
 )
 
+// FormaCommandPersister is responsible for all FormaCommand updates.
+// IMPORTANT: Must remain a singleton actor to ensure serial processing.
```
Comment (JeroenSoeters) [inline internal/metastructure/forma_persister/forma_command_persister.go]: nit: this comment is somewhat redundant. basically everything that is managed by the main supervisor is a singleton, which means we would have to place the comment on all of these actors. probably better to capture this in some future architecture document.

File: internal/api/model/types.go [hunk 5]
```diff
@@ -86,6 +87,18 @@ const (
 	ResourceUpdateStateRejected   = "Rejected"
 )
 
+type TargetUpdate struct {
+	TargetLabel     string    `json:"TargetLabel"`
+	Operation       string    `json:"Operation"`
+	State           string    `json:"State"`
+	Duration        int64     `json:"Duration,omitempty"` // milliseconds
+	ErrorMessage    string    `json:"ErrorMessage,omitempty"`
+	DiscoverableOld bool      `json:"DiscoverableOld,omitempty"`
```
Comment (JeroenSoeters) [inline internal/api/model/types.go]: Don't think this is used anywhere? Can we get away with only a `Discoverable` field?

File: internal/cli/app/app_test.go [hunk 6]
```diff
@@ -3,7 +3,6 @@
 // SPDX-License-Identifier: FSL-1.1-ALv2
 
 //go:build unit
-// +build unit
```
Comment (JeroenSoeters) [inline internal/cli/app/app_test.go:6]: Something went wrong with these build tags, this line got removed in a bunch of test files.

File: internal/metastructure/target_update/target_update.go [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+	"time"
+
+	"github.com/platform-engineering-labs/formae/internal/api/model"
+	"github.com/platform-engineering-labs/formae/internal/metastructure/types"
+	"github.com/platform-engineering-labs/formae/internal/metastructure/util"
+	pkgmodel "github.com/platform-engineering-labs/formae/pkg/model"
+)
+
+type (
+	TargetOperation   = types.OperationType
```
Comment (JeroenSoeters) [inline internal/metastructure/target_update/target_update.go:17]: nit: unrelated to this PR but I noticed the comment on this `OperationType` constant is outdated, we should probably update it that this now pertains to resource updates as well as target updates

File: internal/metastructure/target_update/target_update_generator.go [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+}
+
+// NewTargetUpdateGenerator creates a new target update generator
+func NewTargetUpdateGenerator(ds TargetDatastore) *TargetUpdateGenerator {
+	return &TargetUpdateGenerator{datastore: ds}
+}
+
+// GenerateTargetUpdates determines what target changes are needed
+func (tp *TargetUpdateGenerator) GenerateTargetUpdates(targets []pkgmodel.Target, command pkgmodel.Command) ([]TargetUpdate, error) {
+	// TODO: Generate updates for destroy commands
```
Comment (JeroenSoeters) [inline internal/metastructure/target_update/target_update_generator.go]: nit: let's update this comment so it reflects that we made a conscious decision to delay implementation of this feature and we will support this at some point in the future. now it might come across like something we forgot.

**Review by JeroenSoeters — CHANGES_REQUESTED** [review]
Overall looks good, nice work! Few small comments here and there.

My only concern with this PR is how it affects reconcile, we should add tests for this. The inner query for loading resource modifications since the last reconcile - the one that determines the last reconcile - does not filter on stack: https://github.com/platform-engineering-labs/formae/blob/c20be2913f0f053e1fc2b2436b94e4896f06d908/internal/metastructure/datastore/postgres.go#L528-L532

I'm not 100% sure what the effect of this change is. We should add a test that:
* reconciles a stack
* patches the stack (A)
* creates a target (`--mode reconcile`)
* patches the stack (B)
* reconciles the stack

and ensure that both patch A and patch B show up in the modifications since the last reconcile.

File: internal/api/model/types.go (hunk 5 above)
Comment (browdues) [reply internal/api/model/types.go]: This is used in the display rendering.

File: internal/metastructure/discovery/discovery.go (hunk 3 above)
Comment (browdues) [reply internal/metastructure/discovery/discovery.go]: Since we already warn about deprecation at line 657, and this just confirms the migration completed and tells users to clean up their config, do we need to polish it further given that scanTargets is being removed in the next release anyway (roughly a week away)?

File: internal/metastructure/forma_persister/forma_command_persister.go (hunk 4 above)
Comment (browdues) [reply internal/metastructure/forma_persister/forma_command_persister.go]: Removed

File: internal/cli/app/app_test.go (hunk 6 above)
Comment (browdues) [reply internal/cli/app/app_test.go:6]: This is a golangci-lint issue. Seems like the GH action is pulling latest, and latest may have recently changed. We should consider tying that to a specific version if it becomes a burden.

File: internal/metastructure/target_update/target_update_generator.go (hunk 8 above)
Comment (browdues) [reply internal/metastructure/target_update/target_update_generator.go]: Good call, clarified

File: internal/metastructure/target_update/target_update.go (hunk 7 above)
Comment (browdues) [reply internal/metastructure/target_update/target_update.go:17]: Updated

**Comment by browdues** [comment]
> Overall looks good, nice work! Few small comments here and there.
> 
> My only concern with this PR is how it affects reconcile, we should add tests for this. The inner query for loading resource modifications since the last reconcile - the one that determines the last reconcile - does not filter on stack:
> 
> https://github.com/platform-engineering-labs/formae/blob/c20be2913f0f053e1fc2b2436b94e4896f06d908/internal/metastructure/datastore/postgres.go#L528-L532
> 
> I'm not 100% sure what the effect of this change is. We should add a test that:
> 
> * reconciles a stack
> * patches the stack (A)
> * creates a target (`--mode reconcile`)
> * patches the stack (B)
> * reconciles the stack
> 
> and ensure that both patch A and patch B show up in the modifications since the last reconcile.

Added coverage for this. Thanks for the eyes!

File: internal/metastructure/discovery/discovery.go (hunk 3 above)
Comment (JeroenSoeters) [reply internal/metastructure/discovery/discovery.go]: I'm pretty sure that when I wrote this comment there was a different message here (I believe it said "DEPRECATION WARNING"?). In any event:

>"DEPRECATION WARNING: scan_targets in config file are deprecated. Please use discoverable targets in forma files instead. These config targets will be automatically marked as discoverable.

(1) the config option is `scanTargets`, not `scan_targets`
(2) to me this sentence is a bit vague as it is not directly clear as to what "these config targets" pertains to
(3) in the second message we mention "MIGRATION COMPLETE", but since we don't use the word migration in the first message at all, the user has to guess here that we ran a migration job for their scan targets config, they might wonder if anything else got migrated

I'd say let's print (not log) one clear message to the user what is happening, just print it at the end of the migration, we don't need to spam them that it started and that it subsequently finished since this probably doesn't take more than a second?

File: internal/api/model/types.go (hunk 5 above)
Comment (JeroenSoeters) [reply internal/api/model/types.go]: I cannot find this anywhere?

File: internal/metastructure/discovery/discovery.go (hunk 3 above)
Comment (JeroenSoeters) [reply internal/metastructure/discovery/discovery.go]: FWIW I just started the agent and noticed we have another `INF` message between the 2 warnings, we probably want to get rid of this as well.
